src/core/ext/filters/client_channel/http_connect_handshaker.cc

```diff
@@ -131,28 +130,29 @@ void HttpConnectHandshaker::HandshakeFailedLocked(grpc_error* error) {
 // Callback invoked when finished writing HTTP CONNECT request.
 void HttpConnectHandshaker::OnWriteDone(void* arg, grpc_error* error) {
   auto* handshaker = static_cast<HttpConnectHandshaker*>(arg);
-  gpr_mu_lock(&handshaker->mu_);
+  ReleasableMutexLock lock(&handshaker->mu_);
   if (error != GRPC_ERROR_NONE || handshaker->is_shutdown_) {
     // If the write failed or we're shutting down, clean up and invoke the
     // callback with the error.
     handshaker->HandshakeFailedLocked(GRPC_ERROR_REF(error));
-    gpr_mu_unlock(&handshaker->mu_);
+    lock.Unlock();
     handshaker->Unref();
   } else {
     // Otherwise, read the response.
     // The read callback inherits our ref to the handshaker.
-    grpc_endpoint_read(handshaker->args_->endpoint,
-                       handshaker->args_->read_buffer,
-                       &handshaker->response_read_closure_, /*urgent=*/true);
-    gpr_mu_unlock(&handshaker->mu_);
+    grpc_endpoint* ep = handshaker->args_->endpoint;
```

Comment:
Is it actually necessary to grab these variables before releasing the lock?  I don't think anything else should be accessing them at this point, even if the lock is released.

Same thing on lines 207 and 327.

Reply:
I was mainly worried about CleanupArgsForFailureLocked where we set fields to null within a mutex. This could cause segfault if try to read/write with nullptr fields. Reading the logic again, it looks like we might not hit that, so I think I don't need to grab them. (I'm going to check once more before removing this.)